### PR TITLE
[WIP] Add ctrlp matcher that builds on fruzzy

### PR DESCRIPTION
**Pull request by lervag**
See also #1.

### CODE REVIEW

File: autoload/fruzzy/ctrlp.vim [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  let py = 'pyeval'
+else
+  echom "Can't load fruzzy_ctrlp!"
+  finish
+endif
+
+let s:script_dir = escape(expand('<sfile>:p:h'), '\')
+let s:fruzzy_dir = fnamemodify(s:script_dir, ':h:h') . 'rplugin/python3'
+
+execute pyfile . s:script_dir . '/ctrlp.py'
```
Comment (raghur) [inline autoload/fruzzy/ctrlp.vim]: You could keep the ctrlp.py script in <plugin>/python3 and in the py script, figure out the path to fruzzy.py, add it to sys.path and then import it. Then your vimscript matcher function can just be `import fruzzy_ctrlp; call yourwrapperfunc(args)`

File: autoload/fruzzy/ctrlp.vim [hunk 2]
```diff
@@ -0,0 +1,52 @@
+"
+" Sources:
+" - https://github.com/nixprime/cpsm
+" - https://sourcegraph.com/github.com/nixprime/cpsm@master/-/blob/autoload/cpsm.py#L27
+" - https://github.com/raghur/fruzzy/issues/1
+"
+if has('python3')
+  let pyfile = 'py3file '
+  let py = 'py3eval'
+elseif has('python')
```
Comment (raghur) [inline autoload/fruzzy/ctrlp.vim]: I don't  intend to support py2 - I definitely haven't tested fruzzy or the native extn against it so you can just go with py3

Comment (lervag) [reply autoload/fruzzy/ctrlp.vim]: Great, that makes everything simpler. :)

**Comment by lervag** [comment]
Now it should work. It remains to connect to the nim part. Some questions:

1. There is an extra argument "current" to the fruzzy function. It doesn't seem to do anything. What should I pass to it?

2. The ctrlp matcher has more arguments that could be used:

    ```
    function! fruzzy#ctrlp#matcher(items, str, limit, mmode, ispath, crfile, regex)
      " a:items  : The full list of items to search in.
      " a:str    : The string entered by the user.
      " a:limit  : The max height of the match window. Can be used to limit
      "            the number of items to return.
      " a:mmode  : The match mode. Can be one of these strings:
      "            + "full-line": match the entire line.
      "            + "filename-only": match only the filename.
      "            + "first-non-tab": match until the first tab char.
      "            + "until-last-tab": match until the last tab char.
      " a:ispath : Is 1 when searching in file, buffer, mru, mixed, dir, and
      "            rtscript modes. Is 0 otherwise.
      " a:crfile : The file in the current window. Should be excluded from the
      "            results when a:ispath == 1.
      " a:regex  : In regex mode: 1 or 0.
    ```

    In particular, it seems `a:mmode` should be useful, and possibly also `a:crfile`. What do you think?

**Comment by raghur** [comment]
> There is an extra argument "current" to the fruzzy function. It doesn't seem to do anything. What should I pass to it?

Python version doesn't use that yet. Nim version does something similar to Ctrlp. This is the name of the current buffer

Mmode cannot be used. Crfile is the same as current

**Comment by lervag** [comment]
I'm still working on this. I was trying to get a better highlighting, but I found a strange case. If you run `nvim -u test.vim` in my code, you should get output like this:

```
f
['rplugin/python3/neomru_file', 'rplugin/python3/qc-fast.py', 'rplugin/python3/fruzzy.py', 'plugin/fruzzy.vim', 'autoload/fruzzy.vim']


fm
['rplugin/python3/fruzzy_mod.nim', 'plugin/fruzzy.vim', 'autoload/fruzzy.vim', 'rplugin/python3/denite/filter/matcher/fruzzymatcher.py', 'autoload/fruzzy/ctrlp.vim']


fmf
[]
```

In the last case, with `fmf` as input, I would have expected the result to be `['rplugin/python3/denite/filter/matcher/fruzzymatcher.py']`, but for some reason, it returns an empty list. Is this correct, or did I find a bug?

**Comment by lervag** [comment]
Ok, now things are close to ready. I've settled on the same highlighting method as denite (as you suggested). In this case, it suffices that the fruzzy calls return the list of candidates. The code is very simple.

So far, no options. If the native method is available, it is used, else the pyhon version is used.

What do you think?

Btw: It "feels" slow to parse the candidate list like this:

```python
        return [args['candidates'][i[0]] for i in output]
```

Is there a better way? Perhaps it is not important?

Oh, and finally, right now I'm just passing whatever ctrlp gave for the `current` argument. This should probably be further parsed for fruzzy. How should it be parsed?

**Comment by raghur** [comment]
> So far, no options. If the native method is available, it is used, else the pyhon version is used.

It's trivial to implement. Also I think you should catch ImportError instead of ModuleNotFound

>What do you think?

> Btw: It "feels" slow to parse the candidate list like this:

>        return [args['candidates'][i[0]] for i in output]

I don't know.. as a matter of style, it's feels confusing and hard to understand/parse. I'd suggest
introducing a local variable to make it clearer to read

> Oh, and finally, right now I'm just passing whatever ctrlp gave for the current argument. This should probably be further parsed for fruzzy. How should it be parsed?

So if the list of items is buffer names (ispath=true) and current is one of the buffer names and query is empty, in that case, the list is sorted by shortest levenshtein distance. 
Just make sure that if list of items has absolute filenames then current is also a absolute filename (OTOH, if list has relative names, then current is also relative path)- you'll have to check what ctrlp does in different situations.

**Comment by lervag** [comment]
> > So far, no options. If the native method is available, it is used, else the pyhon version is used.

> It's trivial to implement. ...

Yes, but is there any need to add an option here? That is, if the Nim/native version is available, why not just use it?

In a way, I would think it better if fruzzy provided a single interface to be used by external plugins. Instead of adding a dedicated ctrlp-fruzzy option, there would simply be a single fruzzy option to select the native vs python implementation, e.g. `g:fruzzy_use_native`. This would also simplify the ctrlp wrapper, since it would not need to import more than a single function from fruzzy.

> > Btw: It "feels" slow to parse the candidate list like this:

> >    ```
> >    return [args['candidates'][i[0]] for i in output]
> >    ```

> I don't know.. as a matter of style, it's feels confusing and hard to understand/parse. I'd suggest introducing a local variable to make it clearer to read

Yes, no problem. But my question was related to the use of list comprehensions for fetching the list of matches from the fruzzy results. I could have been more clear. I'm wondering if 1) there is a better/faster way of returning/parsing the list of matches, and/or 2) the scoreMatches function should return the result in a more structured manner. E.g. a dictionary with lists of equal length for the various output types, e.g.:

```
scoreMatchesResult := {
    'matches' : [list of matches],
    'scores' : [list of scores],
    'positions' : [list of positions],
    ...
}
```

In this case, there would only be a single lookup to pass the result back to CtrlP.

> > Oh, and finally, right now I'm just passing whatever ctrlp gave for the current argument. This should probably be further parsed for fruzzy. How should it be parsed?

> So if the list of items is buffer names (ispath=true) and current is one of the buffer names and query is empty, in that case, the list is sorted by shortest levenshtein distance.
> Just make sure that if list of items has absolute filenames then current is also a absolute filename (OTOH, if list has relative names, then current is also relative path)- you'll have to check what ctrlp does in different situations.

Ok, thanks. If I understand correctly:

1. If `ispath` is false, `current` is ignored. Else:
2. If the input list has absolute paths, then `current` should be an absolute path.
2. If the input list has relative paths, then `current` should be a relative path.

**Comment by lervag** [comment]
Ok, I've worked more on this now. In particular, I've tried to get the current keyword to work. I've updated the `test.vim` file, and if you run it, it should produce the following output:

```
Current: ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.vim

Query:
  ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.py
  ~/.vim/bundle/fruzzy/rplugin/python3/fruzzy_mod.nim
  ~/.vim/personal/syntax/dagbok.vim
  ~/.vim/bundle/vimtex/rplugin/python3/denite/source/vimtex.py
  ~/.vim/bundle/fruzzy/rplugin/python3/denite/filter/matcher/fruzzymatcher.py
  ~/.vim/bundle/wiki.vim/pythonx/ncm2_subscope_detector/wiki.py

Query: p
  ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.py
  ~/.vim/bundle/wiki.vim/pythonx/ncm2_subscope_detector/wiki.py
  ~/.vim/bundle/vimtex/rplugin/python3/denite/source/vimtex.py
  ~/.vim/bundle/fruzzy/rplugin/python3/denite/filter/matcher/fruzzymatcher.py
  ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.vim

Query: py
  ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.py
  ~/.vim/bundle/wiki.vim/pythonx/ncm2_subscope_detector/wiki.py
  ~/.vim/bundle/vimtex/rplugin/python3/denite/source/vimtex.py
  ~/.vim/bundle/fruzzy/rplugin/python3/denite/filter/matcher/fruzzymatcher.py
  ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.vim

Query: pym
  ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.py
  ~/.vim/bundle/ctrlp-py-matcher/autoload/pymatcher.vim
  ~/.vim/bundle/fruzzy/rplugin/python3/fruzzy_mod.nim
  ~/.vim/personal/syntax/dagbok.vim
  ~/.vim/bundle/vimtex/rplugin/python3/denite/source/vimtex.py
```

Here I'm sending a list of candidates and a given `current` keyword and increasing queries. As you see, with an empty query, the current file is removed. But for nonempty queries, i.e. when fruzzy is doing the matching, then the current file is not removed.

Is the point that I have to remove the file myself?

I'm pushing a new update where I remove the current file after fruzzy has done the matching, but this feels not quite optimal.

**Comment by lervag** [comment]
> Hmm - I'm only removing the current file if the query's empty. I guess it would make sense to remove it always? Does ctrlp do that?

Yes, in general. For some reason that I can't understand, CtrlP does not remove the file when looking at MRU files. I think this is a bug. IMHO, it makes a lot of sense to remove the current file. And as CtrlP has an option to disable this feature that I've also implemented for the fruzzy matcher, I think it makes a lot of sense to update fruzzy to remove the current file (when supplied and nonempty).

**Comment by lervag** [comment]
> Because it's better to be explicit - esp for testing. I'd hate if I have to move the native version on the fs each time :) just to check the python version.

I don't disagree with "better to be explicit", but I still think it would be better to hide this complexity from the external API. And I don't think this should have any impact on testing. You can still test the two versions independently. You can also add tests for the external/proxy interface, which is harder to do for external users. Both the denite and the ctrlp layer would benefit from this.

But this is of coure up to you. I think the CtrlP layer is close to finished, and my suggestion is only a minor thing.

> Yes - this is g:fruzzy#usenative - I just meant that the ctrlp should respect this config.

Ah, sorry. That makes sense, I'll update. (Again, this wouldn'y be needed if you did add a generic interface.)

> I'll think about it but I'm not too worried about it ATM.

Great. :)

**Comment by raghur** [comment]
Thank you @lervag - Will go through the changes.

BTW, can you add a section to the README on how to use this with CtrlP?

**Comment by lervag** [comment]
I've updated the README file, rebased and cleaned up the commits.

**Comment by raghur** [comment]
@lervag - Thanks a lot for the PR. merged!

**Pull request merged**